Afform - Fix prefilling location blocks with different fields

colemanw opened the pull request:
Overview
----------------------------------------
Fixes [dev/core#5737](https://lab.civicrm.org/dev/core/-/issues/5737) and [dev/core#5738](https://lab.civicrm.org/dev/core/-/issues/5738)

Before
-------
Submitting an afform with multiple location blocks would mess up the location type.

After
------
Fixed.


Technical Details
----------------------------------------
Due to the way joins were originally designed, the multiple blocks aren't treated separately. So any fields in one are accepted for all. This fixes it so the last block's list of fields don't override the others, and fixes the location types all getting jumbled together.

colemanw committed:
Afform - Fix prefilling location blocks with different fields

Fixes dev/core#5737

colemanw committed:
FormBuilder - Fix saving multiple location blocks

Fixes dev/core#5738

samuelsov commented:
@colemanw The prefill seems good and saving existing data work as well but adding data seems broken:
- with the same Formbuilder definition
- take a contact with no phone and add phone(s) using the Formbuilder
- => Phones are not updated

colemanw commented:
@samuelsov thanks for testing this. I added more tests to cover the issue you described, but the tests all pass -- adding a new phone and/or updating an existing one all seems to work fine.

samuelsov commented:
Odd, my tests are on the online demo of the PR http://core-32121-83n77.test-1.civicrm.org:8003/civicrm/pr32121/#?Individual1=126... I'll try to reproduce on a dev site

colemanw committed:
FormBuilder - Fix saving new location blocks

When an existing contact doesn't have a record for a particular location,
(e.g. work phone), it should be an empty object,
which unfortunately gets mangled by `json_encode` and ends up being
an empty array.

Switching to `null` works better.

colemanw commented:
@samuelsov I was able to reproduce the bug you mentioned and just pushed up a fix to this PR.
